geps/gep-1731/index.md

```diff
@@ -0,0 +1,409 @@
+    // response from the backend, the Gateway MUST return an error.
+    //
+    // When this field is unspecified, the number of times to attempt to retry
+    // a backend request is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Attempts *Int `json:"attempts,omitempty"`
+
+    // Backoff specifies the minimum duration a Gateway should wait between
+    // retry attempts and is represented in Gateway API Duration formatting.
+    //
+    // For example, setting the `rules[].retry.backoff` field to the value
+    // `100ms` will cause a backend request to first be retried approximately
+    // 100 milliseconds after timing out or reciveing a response code configured
+    // to be retryable.
+    //
+    // An implementation MAY use an exponential or alternative backoff strategy
+    // for subsequent retry attempts, MAY cap the maximum backoff duration to
+    // some amount greater than the specified minimum, and MAY add arbitrary
+    // jitter to stagger requests, as long as unsuccessful backend requests are
+    // not retried before the configured minimum duration.
+    //
+    // If a Request timeout (`rules[].timeouts.request`) is configured on the
+    // route, the entire duration of the initial request and any retry attempts
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching any
+// status code in the 500-599 range, and SHOULD support specifying any
+// status code within the valid 100-599 (inclusive) range.
```

Comment:
100-599 as an inclusive range seems too broad. What happens if someone wants to retry on 200 statuses? I think it would be safer to leave out any suggestion of support for status codes <400.

Reply:
(See above, I'd vote for SHOULD NOT (but MAY) 100-499.)

Reply:
Is retrying 429s really a bad idea? That code in particular seems like it _should_ be retried, just with some additional logic/delay that I didn't want to get into here yet (and could possibly be part of a GEP for https://github.com/kubernetes-sigs/gateway-api/issues/326 at some point in the future).

Reply:
Added `SHOULD NOT support 100-399` and `MAY support 400-499` with a warning, consolidating this discussion into thread above.